src/celeritas/phys/RootEventSampler.cc

```diff
@@ -0,0 +1,63 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/phys/RootEventSampler.cc
+//---------------------------------------------------------------------------//
+#include "RootEventSampler.hh"
+
+#include <algorithm>
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct \c RootEventReader , and initialize sampling conditions.
+ */
+RootEventSampler::RootEventSampler(std::string const& filename,
+                                   SPConstParticles params,
+                                   size_type num_clumped_events,
+                                   unsigned int seed)
+    : num_clumped_events_(num_clumped_events)
+{
+    CELER_EXPECT(!filename.empty());
+    CELER_EXPECT(params);
+
+    reader_ = std::make_unique<RootEventReader>(filename, std::move(params));
+    CELER_EXPECT(num_clumped_events_ > 0
+                 && num_clumped_events_ <= reader_->num_events());
+
+    rng_.seed(seed);
+    select_event_
+        = std::uniform_int_distribution<std::size_t>(0, reader_->num_events());
```

Comment:
Should we be sampling events without replacement?

Reply:
I think replacement is fine for the intended purpose

Reply:
Discussed with @stognini offline: this should sample with replacement but not return empty events after the number of events in the root file have been sampled.